### ISSUE TITLE
seg faults in usapt0 solver 

### ISSUE THREAD

**Issue by PhillCli** [psi4/psi4#1830]
Hi,
for odd cases where there is no electron in either alpha or beta block, sapt0 solver for UHF will result in segfault due to ```zero``` dimension passed to some of the procedures, e.g. ```C_DGEMM``` calls.

minimal working example
```Python
import psi4

psi4.set_memory("1024 MB")
geo_string = """
0 2
Li 0.000 0.000 -1.000
--
0 2
H  0.000 0.000 +1.000

no_reorient
no_com
units angstrom
symmetry c1
""".strip()

BASIS = "aug-cc-pvdz"
REFERENCE = "uhf"
BASIS_GUESS = "3-21G"
GUESS = "sad"
dimer = psi4.geometry(geo_string)

psi4.set_options(
    {
        "basis": BASIS,
        "reference": REFERENCE,
        "basis_guess": BASIS_GUESS,
        "guess": GUESS,
        "scf_type": "mem_df",
        "e_convergence": 1e-12,
        "d_convergence": 1e-12,
    }
)

psi4.energy("sapt0", return_wfn=True, molecule=dimer)
```

result:
```
    Role: ORBITAL
    Keyword: BASIS
    atoms 1 entry LI         line    35 file /home/filip/software/psi4-1.4rc/share/psi4/basis/3-21g.gbs 
    atoms 2 entry H          line    21 file /home/filip/software/psi4-1.4rc/share/psi4/basis/3-21g.gbs 


         ---------------------------------------------------------
                           SCF Castup computation                  
         ---------------------------------------------------------
                                   SCF
               by Justin Turney, Rob Parrish, Andy Simmonett
                          and Daniel G. A. Smith
                              UHF Reference
                        1 Threads,    976 MiB Core
         ---------------------------------------------------------

  ==> Geometry <==

    Molecular point group: c1
    Full point group: C_inf_v

    Geometry (in Angstrom), charge = 0, multiplicity = 2:

       Center              X                  Y                   Z               Mass       
    ------------   -----------------  -----------------  -----------------  -----------------
      Gh(LI)          0.000000000000     0.000000000000    -1.000000000000     7.016003436600
         H            0.000000000000     0.000000000000     1.000000000000     1.007825032230

  Running in c1 symmetry.

  Rotational constants: A = ************  B =      2.10095  C =      2.10095 [cm^-1]
  Rotational constants: A = ************  B =  62984.77227  C =  62984.77227 [MHz]
  Nuclear repulsion =    0.000000000000000

  Charge       = 0
  Multiplicity = 2
  Electrons    = 1
  Nalpha       = 1
  Nbeta        = 0

  ==> Algorithm <==

  SCF Algorithm Type is DF.
  DIIS enabled.
  MOM disabled.
  Fractional occupation disabled.
  Guess Type is SAD.
  Energy threshold   = 1.00e-12
  Density threshold  = 1.00e-12
  Integral threshold = 0.00e+00

  ==> Primary Basis <==

  Basis Set: 3-21G
    Blend: 3-21G
    Number of shells: 7
    Number of basis function: 11
    Number of Cartesian functions: 11
    Spherical Harmonics?: false
    Max angular momentum: 1

   => Loading Basis Set <=

    Name: (3-21G AUX)
    Role: JKFIT
    Keyword: DF_BASIS_SCF
    atoms 1 entry LI         line    59 file /home/filip/software/psi4-1.4rc/share/psi4/basis/def2-svp-jkfit.gbs 
    atoms 2 entry H          line    23 file /home/filip/software/psi4-1.4rc/share/psi4/basis/def2-svp-jkfit.gbs 

  ==> Pre-Iterations <==

   -------------------------------------------------------
    Irrep   Nso     Nmo     Nalpha   Nbeta   Ndocc  Nsocc
   -------------------------------------------------------
     A         11      11       0       0       0       0
   -------------------------------------------------------
    Total      11      11       1       0       0       1
   -------------------------------------------------------

  ==> Integral Setup <==

  ==> DiskDFJK: Density-Fitted J/K Matrices <==

    J tasked:                  Yes
    K tasked:                  Yes
    wK tasked:                  No
    OpenMP threads:              1
    Integrals threads:           1
    Memory [MiB]:              732
    Algorithm:                Core
    Integral Cache:           NONE
    Schwarz Cutoff:          1E-12
    Fitting Condition:       1E-10

   => Auxiliary Basis Set <=

  Basis Set: (3-21G AUX)
    Blend: DEF2-SVP-JKFIT
    Number of shells: 23
    Number of basis function: 80
    Number of Cartesian functions: 80
    Spherical Harmonics?: false
    Max angular momentum: 3

  Minimum eigenvalue in the overlap matrix is 1.9324024167E-01.
  Using Symmetric Orthogonalization.

  SCF Guess: Superposition of Atomic Densities via on-the-fly atomic UHF.

  ==> Iterations <==

                           Total Energy        Delta E     RMS |[F,P]|

   @DF-UHF iter SAD:    -0.34299486696379   -3.42995e-01   0.00000e+00 
   @DF-UHF iter   1:    -0.48778126858064   -1.44786e-01   7.18202e-03 DIIS
   @DF-UHF iter   2:    -0.49627418123381   -8.49291e-03   2.58017e-03 DIIS
   @DF-UHF iter   3:    -0.49762079007351   -1.34661e-03   1.09982e-04 DIIS
   @DF-UHF iter   4:    -0.49762397190176   -3.18183e-06   1.60614e-05 DIIS
   @DF-UHF iter   5:    -0.49762403062183   -5.87201e-08   1.47415e-06 DIIS
   @DF-UHF iter   6:    -0.49762403115935   -5.37526e-10   6.69153e-08 DIIS
   @DF-UHF iter   7:    -0.49762403116031   -9.60343e-13   2.69685e-09 DIIS
   @DF-UHF iter   8:    -0.49762403116031   -7.77156e-16   2.10910e-10 DIIS
   @DF-UHF iter   9:    -0.49762403116032   -8.88178e-16   2.91582e-11 DIIS
   @DF-UHF iter  10:    -0.49762403116032   -6.66134e-16   1.10373e-14 DIIS
  Energy and wave function converged.


  ==> Post-Iterations <==

   @Spin Contamination Metric:   0.000000000E+00
   @S^2 Expected:                7.500000000E-01
   @S^2 Observed:                7.500000000E-01
   @S   Expected:                5.000000000E-01
   @S   Observed:                5.000000000E-01

    Orbital Energies [Eh]
    ---------------------

    Alpha Occupied:                                                       

       1A     -0.497624  

    Alpha Virtual:                                                        

       2A      0.048719     3A      0.066612     4A      0.066612  
       5A      0.098792     6A      0.253872     7A      0.387743  
       8A      0.387743     9A      0.427498    10A      1.213680  
      11A      3.852115  

    Beta Occupied:                                                        

    

    Beta Virtual:                                                         

       1A      0.020894     2A      0.058417     3A      0.068354  
       4A      0.068354     5A      0.183800     6A      0.269157  
       7A      0.391375     8A      0.391375     9A      0.451441  
      10A      1.378124    11A      3.853058  

    Final Occupation by Irrep:
              A 
    DOCC [     0 ]
    SOCC [     1 ]

  @DF-UHF Final Energy:    -0.49762403116032

   => Energetics <=

    Nuclear Repulsion Energy =              0.0000000000000000
    One-Electron Energy =                  -0.4976240311603163
    Two-Electron Energy =                   0.0000000000000000
    Total Energy =                         -0.4976240311603163

  UHF NO Occupations:
  HONO-0 :    1  A 1.0000000
  LUNO+0 :    2  A 0.0000000
  LUNO+1 :    3  A 0.0000000
  LUNO+2 :    4  A 0.0000000
  LUNO+3 :    5  A 0.0000000


Computation Completed

  //>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>//
  //                SCF                //
  //<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<//

   => Loading Basis Set <=

    Name: AUG-CC-PVDZ
    Role: ORBITAL
    Keyword: BASIS
    atoms 1 entry LI         line    68 file /home/filip/software/psi4-1.4rc/share/psi4/basis/aug-cc-pvdz.gbs 
    atoms 2 entry H          line    40 file /home/filip/software/psi4-1.4rc/share/psi4/basis/aug-cc-pvdz.gbs 


         ---------------------------------------------------------
                                   SCF
               by Justin Turney, Rob Parrish, Andy Simmonett
                          and Daniel G. A. Smith
                              UHF Reference
                        1 Threads,    976 MiB Core
         ---------------------------------------------------------

  ==> Geometry <==

    Molecular point group: c1
    Full point group: C_inf_v

    Geometry (in Angstrom), charge = 0, multiplicity = 2:

       Center              X                  Y                   Z               Mass       
    ------------   -----------------  -----------------  -----------------  -----------------
      Gh(LI)          0.000000000000     0.000000000000    -1.000000000000     7.016003436600
         H            0.000000000000     0.000000000000     1.000000000000     1.007825032230

  Running in c1 symmetry.

  Rotational constants: A = ************  B =      2.10095  C =      2.10095 [cm^-1]
  Rotational constants: A = ************  B =  62984.77227  C =  62984.77227 [MHz]
  Nuclear repulsion =    0.000000000000000

  Charge       = 0
  Multiplicity = 2
  Electrons    = 1
  Nalpha       = 1
  Nbeta        = 0

  ==> Algorithm <==

  SCF Algorithm Type is MEM_DF.
  DIIS enabled.
  MOM disabled.
  Fractional occupation disabled.
  Guess Type is SAD.
  Energy threshold   = 1.00e-12
  Density threshold  = 1.00e-12
  Integral threshold = 0.00e+00

  ==> Primary Basis <==

  Basis Set: AUG-CC-PVDZ
    Blend: AUG-CC-PVDZ
    Number of shells: 14
    Number of basis function: 32
    Number of Cartesian functions: 34
    Spherical Harmonics?: true
    Max angular momentum: 2

   => Loading Basis Set <=

    Name: (AUG-CC-PVDZ AUX)
    Role: JKFIT
    Keyword: DF_BASIS_SCF
    atoms 1 entry LI         line    59 file /home/filip/software/psi4-1.4rc/share/psi4/basis/def2-qzvpp-jkfit.gbs 
    atoms 2 entry H          line    70 file /home/filip/software/psi4-1.4rc/share/psi4/basis/aug-cc-pvdz-jkfit.gbs 


  Computing basis projection from 3-21G to AUG-CC-PVDZ

  ==> Pre-Iterations <==

   -------------------------------------------------------
    Irrep   Nso     Nmo     Nalpha   Nbeta   Ndocc  Nsocc
   -------------------------------------------------------
     A         32      32       0       0       0       0
   -------------------------------------------------------
    Total      32      32       1       0       0       1
   -------------------------------------------------------

  ==> Integral Setup <==

  DFHelper Memory: AOs need 0.001 GiB; user supplied 0.715 GiB. Using in-core AOs.

  ==> MemDFJK: Density-Fitted J/K Matrices <==

    J tasked:                   Yes
    K tasked:                   Yes
    wK tasked:                   No
    OpenMP threads:               1
    Memory [MiB]:               732
    Algorithm:                 Core
    Schwarz Cutoff:           1E-12
    Mask sparsity (%):       0.0000
    Fitting Condition:        1E-10

   => Auxiliary Basis Set <=

  Basis Set: (AUG-CC-PVDZ AUX)
    Blend: AUG-CC-PVDZ-JKFIT + DEF2-QZVPP-JKFIT
    Number of shells: 29
    Number of basis function: 83
    Number of Cartesian functions: 95
    Spherical Harmonics?: true
    Max angular momentum: 3

  Minimum eigenvalue in the overlap matrix is 3.7125823143E-03.
  Using Symmetric Orthogonalization.

  SCF Guess: Orbitals guess was supplied from a previous computation.

  ==> Iterations <==

                           Total Energy        Delta E     RMS |[F,P]|

   @DF-UHF iter   0:    -0.49904529387990   -4.99045e-01   5.81466e-04 
   @DF-UHF iter   1:    -0.49933835527994   -2.93061e-04   9.46849e-05 DIIS
   @DF-UHF iter   2:    -0.49935062691594   -1.22716e-05   2.23284e-05 DIIS
   @DF-UHF iter   3:    -0.49935159814175   -9.71226e-07   4.41164e-06 DIIS
   @DF-UHF iter   4:    -0.49935164375895   -4.56172e-08   5.33254e-07 DIIS
   @DF-UHF iter   5:    -0.49935164423349   -4.74539e-10   5.00344e-08 DIIS
   @DF-UHF iter   6:    -0.49935164423748   -3.99308e-12   6.43193e-09 DIIS
   @DF-UHF iter   7:    -0.49935164423756   -8.04357e-14   8.07626e-10 DIIS
   @DF-UHF iter   8:    -0.49935164423756   -5.55112e-16   3.13992e-11 DIIS
   @DF-UHF iter   9:    -0.49935164423756   -5.55112e-17   4.79952e-12 DIIS
   @DF-UHF iter  10:    -0.49935164423756    7.21645e-16   2.32387e-13 DIIS
  Energy and wave function converged.


  ==> Post-Iterations <==

   @Spin Contamination Metric:   0.000000000E+00
   @S^2 Expected:                7.500000000E-01
   @S^2 Observed:                7.500000000E-01
   @S   Expected:                5.000000000E-01
   @S   Observed:                5.000000000E-01

    Orbital Energies [Eh]
    ---------------------

    Alpha Occupied:                                                       

       1A     -0.499352  

    Alpha Virtual:                                                        

       2A      0.012936     3A      0.013757     4A      0.013757  
       5A      0.015378     6A      0.063621     7A      0.068900  
       8A      0.068900     9A      0.086764    10A      0.103920  
      11A      0.198954    12A      0.198954    13A      0.208977  
      14A      0.208977    15A      0.256823    16A      0.270230  
      17A      0.318044    18A      0.318044    19A      0.402166  
      20A      0.450346    21A      0.450346    22A      0.540514  
      23A      0.606544    24A      0.606544    25A      0.613724  
      26A      0.613724    27A      0.838452    28A      1.281252  
      29A      1.826605    30A      1.826605    31A      1.850798  
      32A      4.179388  

    Beta Occupied:                                                        

    

    Beta Virtual:                                                         

       1A      0.006309     2A      0.013514     3A      0.013821  
       4A      0.013821     5A      0.037931     6A      0.065570  
       7A      0.070561     8A      0.070561     9A      0.090473  
      10A      0.171679    11A      0.207466    12A      0.207466  
      13A      0.209492    14A      0.209492    15A      0.257933  
      16A      0.287508    17A      0.331720    18A      0.331720  
      19A      0.427501    20A      0.478623    21A      0.478623  
      22A      0.598357    23A      0.612674    24A      0.612674  
      25A      0.614564    26A      0.614564    27A      0.921140  
      28A      1.322779    29A      1.926318    30A      1.926318  
      31A      1.947876    32A      4.183965  

    Final Occupation by Irrep:
              A 
    DOCC [     0 ]
    SOCC [     1 ]

  @DF-UHF Final Energy:    -0.49935164423756

   => Energetics <=

    Nuclear Repulsion Energy =              0.0000000000000000
    One-Electron Energy =                  -0.4993516442375593
    Two-Electron Energy =                   0.0000000000000000
    Total Energy =                         -0.4993516442375593

  UHF NO Occupations:
  HONO-0 :    1  A 1.0000000
  LUNO+0 :    2  A 0.0000000
  LUNO+1 :    3  A 0.0000000
  LUNO+2 :    4  A 0.0000000
  LUNO+3 :    5  A 0.0000000


Computation Completed


Properties will be evaluated at   0.000000,   0.000000,   0.000000 [a0]

Properties computed using the SCF density matrix

  Nuclear Dipole Moment: [e a0]
     X:     0.0000      Y:     0.0000      Z:     1.8897

  Electronic Dipole Moment: [e a0]
     X:     0.0000      Y:    -0.0000      Z:    -1.8895

  Dipole Moment: [e a0]
     X:     0.0000      Y:    -0.0000      Z:     0.0002     Total:     0.0002

  Dipole Moment: [D]
     X:     0.0000      Y:    -0.0000      Z:     0.0006     Total:     0.0006


*** tstop() called on filip-ThinkPad-E495 at Wed Mar  4 16:29:33 2020
Module time:
	user time   =       0.55 seconds =       0.01 minutes
	system time =       0.01 seconds =       0.00 minutes
	total time  =          0 seconds =       0.00 minutes
Total time:
	user time   =       1.79 seconds =       0.03 minutes
	system time =       0.07 seconds =       0.00 minutes
	total time  =          2 seconds =       0.03 minutes
  Constructing Basis Sets for SAPT...

   => Loading Basis Set <=

    Name: (AUG-CC-PVDZ AUX)
    Role: RIFIT
    Keyword: DF_BASIS_SAPT
    atoms 1 entry LI         line    90 file /home/filip/software/psi4-1.4rc/share/psi4/basis/def2-qzvpp-ri.gbs 
    atoms 2 entry H          line    30 file /home/filip/software/psi4-1.4rc/share/psi4/basis/aug-cc-pvdz-ri.gbs 


  //>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>//
  //               SAPT0               //
  //<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<//


*** tstart() called on filip-ThinkPad-E495
*** at Wed Mar  4 16:29:34 2020

	 --------------------------------------------------------
	                         SAPT                      
	               Rob Parrish and Ed Hohenstein             
	                Open-shell: Jérôme Gonthier              
	 --------------------------------------------------------

  ==> Sizes <==

   => Resources <=

    Memory [MiB]:              878

   => Orbital Ranges <=

    ------------------
    Range    M_A   M_B
    ------------------
    natom      1     1
    nso       32    32
    ------------------
      Alpha orbitals  
    ------------------
    nmo       32    32
    nocc       2     1
    nvir      30    31
    nfocc      0     0
    naocc      2     1
    navir     30    31
    nfvir      0     0
    ------------------
      Beta orbitals  
    ------------------
    nmo       32    32
    nocc       1     0
    nvir      31    32
    nfocc      0     0
    naocc      1     0
    navir     31    32
    nfvir      0     0
    ------------------

   => Primary Basis Set <=

  Basis Set: AUG-CC-PVDZ
    Blend: AUG-CC-PVDZ
    Number of shells: 14
    Number of basis function: 32
    Number of Cartesian functions: 34
    Spherical Harmonics?: true
    Max angular momentum: 2

  SCF TERMS:

[1]    11021 segmentation fault (core dumped)  python minimal_example.py
```

```
Program received signal SIGSEGV, Segmentation fault.
0x00007fffea9b8e06 in psi::sapt::USAPT0::build_Sijb(std::shared_ptr<psi::Matrix>) ()
   from /home/filip/software/psi4-1.4rc/lib/psi4/core.cpython-36m-x86_64-linux-gnu.so
```

**Comment by jgonthier** [comment]
I am attempting to work on this issue. Problem is, I have very little time so it may take some time. Also, I have not touched the Psi4 code in 3 years. @loriab is there a state-of-the-art way to debug Psi4? I found some slides by Justin Turney from 2017, but I couldn't make gdb work for now.

**Comment by loriab** [comment]
Thanks, @jgonthier. The gdb procedure these days is `gdb --args python stage/bin/psi4 input.dat`. It should work readily -- I've used it this year. Please ping me if it gives trouble. I think @zachglick is also interested. http://psicode.org/psi4manual/master/build_planning.html#how-to-use-gdb-and-lldb-with-psi4

**Comment by jgonthier** [comment]
Little update: looking at formulas, I think we should just set a bunch of exchange terms involving the absent spin to 0. Probably check number of alpha and beta electrons of each monomer at the beginning of the code and then select which terms to compute based on that. I'm set up with debugging (thanks Lori!) so I'll try that soon.

**Comment by jgonthier** [comment]
Started fix in [USAPT_spin_fix](https://github.com/jgonthier/psi4/tree/USAPT_spin_fix) branch on my fork.

**Comment by PhillCli** [comment]
@loriab 
as a follow up from #1948
the original issue was resolved, but I've missed another edge-case, where setting `freeze_core True` results in empty-blocks in mp2 terms ... (sorry for being such a pain).
I've hacked away a solution and additional regression tests (should I add `output.ref` also?) with lithium dimer.
The code is available at [fcore_fix](https://github.com/PhillCli/psi4/tree/usapt0_spin_fix_fcore), should I make another PR?

**Comment by loriab** [comment]
@PhillCli 

> as a follow up from #1948 the original issue was resolved, but I've missed another edge-case, where setting freeze_core True results in empty-blocks in mp2 terms ... (sorry for being such a pain).
I've hacked away a solution and additional regression tests (should I add output.ref also?) with lithium dimer.

Not at all -- problem identified and solution prepared is a double win, not a pain. thank you.

> The code is available at fcore_fix, should I make another PR?

please do. and `output.ref` welcome. we try to generate missing ones every year or so, but sooner they're present the better.

**Comment by hokru** [comment]
resolved by #1969
